### PR TITLE
Spark 3.5: Don't cache or reuse manifest entries while rewriting metadata by default

### PR DESCRIPTION
**Pull request by aokolnychyi**
The action for rewriting manifests caches the manifest entry DF or does an extra shuffle in order to skip reading the actual manifest files twice. We did this assuming it would increase the performance. However, the caching seems to perform poorly for larger tables as it requires substantial cluster resources. In addition, doing a round-robin repartition is expensive as the entries must be written to disk. The extra write is actually more expensive than the extra read required for the range-based shuffle of manifest entries.

Therefore, this PR disables caching by default and removes the optional round-robin repartition step. Instead, we will read the manifests twice (this step is distributed and scales really well even for tables with huge metadata). I did some tests on a cluster and saw operations completed 2x faster. Moreover, there were some cases when caching caused failures. The new approach should be both faster and more robust.

### CODE REVIEW

File: spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteManifestsSparkAction.java [hunk 1]
```diff
@@ -277,16 +275,9 @@ private List<ManifestFile> writeManifestsForPartitionedTable(
...
     boolean useCaching =
         PropertyUtil.propertyAsBoolean(options(), USE_CACHING, USE_CACHING_DEFAULT);
-    if (useCaching) {
-      reusableDS = ds.cache();
-    } else {
-      int parallelism = SQLConf.get().numShufflePartitions();
-      reusableDS =
-          ds.repartition(parallelism).map((MapFunction<T, T>) value -> value, ds.exprEnc());
-    }
+    Dataset<T> reusableDS = useCaching ? ds.cache() : ds;
```
Comment (RussellSpitzer) [inline spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteManifestsSparkAction.java:280]: The repartition is also removed here, is that intentionally?

**Review by RussellSpitzer — APPROVED** [review]
I think this is a much safer default than our current, Caching is not my favorite feature of Spark and it is unstable sometimes. I do want to know why the shuffle is being removed.

File: spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteManifestsSparkAction.java (hunk 1 above)
Comment (RussellSpitzer) [reply spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteManifestsSparkAction.java:280]: Ah I see your note in the PR

Comment (aokolnychyi) [reply spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteManifestsSparkAction.java:280]: Yeah, I found that write to be more expensive than an extra read.

**Comment by aokolnychyi** [comment]
Thanks, @RussellSpitzer!

**Pull request merged**